### ISSUE TITLE
Using Polylang Pro with composer

### ISSUE THREAD

**Issue by alexkerber** [polylang/polylang#3]
Any Ideas?

**Comment by grappler** [comment]
I use [Release Belt](https://github.com/Rarst/release-belt) for all of the commercial plugins that I buy.

Or you could use the download link like explained here https://deliciousbrains.com/wp-migrate-db-pro/doc/installing-via-composer/

**Comment by alexkerber** [comment]
Problem is that with pro the download link will expire after some time/downloads

**Comment by mcaskill** [comment]
A Composer manifest for the free version would also be nice.

**Comment by grappler** [comment]
You can use https://wpackagist.org, After adding it as a repository, run `composer require wpackagist-plugin/polylang`

**Comment by alexkerber** [comment]
yes wpackagist works perfect without pro

**Comment by Chouby** [comment]
I rely on Easy Digital Downloads for downloads and license management. And indeed the problem is that it generates a dynamic link. So I fear that it's not possible for now.

**Comment by intelligence** [comment]
Just wanted to add that I'd like to see this in the future aswell!

**Comment by E-VANCE** [comment]
Same here, would appreciate to find a way to integrate Polylang Pro into my Composer workflow

**Comment by alfhen** [comment]
Yeah I am also very interested in having this feature, a solution like @grappler suggested like https://deliciousbrains.com/wp-migrate-db-pro/doc/installing-via-composer/ would be much appreciated.

**Comment by rask** [comment]
Any news on this? WPackagist is awesome but 99% of paid plugins are currently causing headaches for developers and ops. Would it help to contact EDD authors with a feature request regarding more static download URLs?

**Comment by nlemoine** [comment]
https://github.com/junaidbhura/composer-wp-pro-plugins

Just found this but the install process keeps asking me a username and password.

It would be really nice to have a composer install option.

**Comment by rubas** [comment]
*Bump* for 2020 - we have working solutions for ACF, AdminColumns, Yoast .. like to see one for Polylang Pro too.

**Comment by alexkerber** [comment]
@rubas any instructions? :)

**Comment by rubas** [comment]
- ACF & AdminColumns have not dynamic download links, just a version and private key. You can use [ffraenz/private-composer-installer](https://github.com/ffraenz/private-composer-installer) for it.
- Yoast & Deliciousbrains runs their own repositiory under https://my.yoast.com/packages/ and protect it via an http-auth.

- For Polylang Po, you have to roll your own ie copy it to private git and distribute directly or via toran proxy/private packagist.

**Comment by alexkerber** [comment]
I'll test this during the weekend: https://github.com/junaidbhura/composer-wp-pro-plugins

**Comment by rubas** [comment]
I [see](https://github.com/junaidbhura/composer-wp-pro-plugins/blob/master/plugins/PolylangPro.php) -  I wasn't aware of this. Thanks.

**Comment by ajitbohra** [comment]
The following worked for me:

1. Follow the instructions here https://github.com/junaidbhura/composer-wp-pro-plugins#usage
2. update `.env` to include
```
POLYLANG_PRO_KEY='yourkey'
POLYLANG_PRO_URL='yourdomain'
```
3. use composer `auth.json` https://getcomposer.org/doc/articles/http-basic-authentication.md
```
{
  "http-basic": {
    "www.polylang.pro": {
      "username": "yourusername",
      "password": "yourpassword"
    }
  }
}
```

**Comment by szepeviktor** [comment]
https://github.com/ffraenz/private-composer-installer is a superior product
Please consider using it.

**Comment by ajitbohra** [comment]
@szepeviktor maybe I am missing something can you help with the steps to make `polylang-pro` work with `ffraenz/private-composer-installer`?

**Comment by szepeviktor** [comment]
I do not know how it works. But `private-composer-installer`˙is superior as it is universal.

I see polylang-pro is distributed by EDD
https://github.com/junaidbhura/composer-wp-pro-plugins/blob/master/plugins/PolylangPro.php#L38-L51
So you need to POST once and receive a JSON object with the download URL.

**Comment by szepeviktor** [comment]
I've found the docs!! :)
https://docs.easydigitaldownloads.com/article/384-software-licensing-api

**Comment by ajitbohra** [comment]
`ffraenz/private-composer-installer` works well when we have a permanent direct URL for download. In the case of `polylang pro` we don't have that, download URL expire after a certain time and need to be fetched via post request to API.

**Comment by szepeviktor** [comment]
Yes.
One should develop an EDD Software Licensing Composer Plugin :)

e.g. https://packagist.org/packages/szepeviktor/composer-envato

**Comment by ajitbohra** [comment]
@szepeviktor have developed the quick package `Composer EDD Plugin' to install any plugin powered by EDD SL Addon.

Planning to release it next week.

**Comment by ajitbohra** [comment]
Composer plugin for EDD
https://github.com/lubusIN/composer-edd-plugin

**Comment by przyb** [comment]
It's 2024, Composer now offers an official way to handle private repos with `auth.json` (ACF now switched to using that), and we are still stuck without an official support for Polylang Pro. C'mon, it's time to give your clients some love. It's getting harder and harder to recommend PLL Pro to our clients when the DX is so poor.

**Comment by szepeviktor** [comment]
@Chouby can do that. I know.

**Comment by clrblnd** [comment]
Adding another request for composer to the bucket 📈

**Comment by E-VANCE** [comment]
In the meantime you might want to take a look at [SatisPress](https://github.com/cedaro/satispress), which allows managing & installing plugins like Polylang PRO via Composer.

**Comment by hirasso** [comment]
Thanks @E-VANCE for the tip. 

Instead of yet another tool, I decided to check-in a `polylang-pro.zip` in my project's (private) git repo. Now I can automatically install polylang-pro from that zip file every time `composer install` runs. In my `composer.json`:


```json
{
  "name": "rah/wp-boilerplate",
  "scripts": {
    "post-install-cmd": [
      "@install-polylang-pro"
    ],
    "install-polylang-pro": [
      "rm -rf ./public/content/plugins/polylang-pro",
      "wp plugin install ./polylang-pro.zip --activate --skip-plugins"
    ],
  }
}

```
`./public/content/plugins` is where I have my WordPress plugins stored. It's a slightly modified roots/bedrock folder structure.

That way, my site stays installable, even if I get hit by a bus ;)

**Comment by hirasso** [comment]
For what it's worth – here is the response to my support request:

> Hello,
> 
> The feature you're requesting is one of the oldest from our customers, and it's always in our todolist to add it, but we worked on others features for the moment.
Adding this feature is not only plugin related, we have to set up a server which will deliver the packages. It also requires some works with our host provider.
As of today, we always suggest to use https://github.com/junaidbhura/composer-wp-pro-plugins to install Polylang Pro through Composer.
> 
> Using the native WordPress way to updates plugins should not break the confidence you have in Polylang Pro. We're maintaining our products and improving it regularly.
Polylang Pro 3.7 is on its way to bring to you some great improvements. https://polylang.pro/polylang-pro-3-7-beta1-is-now-available-for-download/
>
> Regards

That does not read to me as they are planning to add this feature anytime soon 😞

**Comment by hirasso** [comment]
@E-VANCE sorry I was so frustrated with polylang and composer that I hacked my solution together without really looking at SatisPress. I promise I'll "give it 5 minutes" tomorrow. Thank you for the tip!! ♥️

**Comment by hirasso** [comment]
I've ended up setting up a satispress instance to add composer support for my premium plugins. It's amazing!! Huge thanks again to @E-VANCE for this tip!

**Comment by dvaansiseelr** [comment]
Still no updates? it's been 10 years already since the first issue.

**Comment by szepeviktor** [comment]
@dvaansiseelr I've made this Composer plugin for you. https://github.com/szepeviktor/Chouby-decade

**Comment by dvaansiseelr** [comment]
> [@dvaansiseelr](https://github.com/dvaansiseelr) I've made this Composer plugin for you. https://github.com/szepeviktor/Chouby-decade

Thank you for building this — I love it very much. I assume it's designed for a traditional WordPress environment. In a WP Bedrock environment, I'd like to add the PRO license key in the .env file instead of running:
`POLYLANG_PRO_LICENSE_KEY='your-license-key' composer require wpsyntex/polylang-pro:^3.8,`
Unfortunately, running `composer update` doesn't let me install it this way.
'The Polylang Pro API did not return a secure package download URL. Check the license and its activation.'
